File: web/validation.py
"""
Input validation utilities for web interface
Provides validation for domains, IPs, CIDR notation, and other user inputs
"""

import re
import ipaddress
from typing import List, Tuple

def validate_domain(domain: str) -> Tuple[bool, str]:
    """
    Validate domain name format
    
    Args:
        domain: Domain name to validate
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not domain or not isinstance(domain, str):
        return False, "Domain cannot be empty"
    
    # Remove whitespace
    domain = domain.strip()
    
    # Check length
    if len(domain) > 253:
        return False, "Domain name too long (max 253 characters)"
    
    # Domain regex pattern
    # Allows alphanumeric, hyphens, dots, and underscores (for AD)
    pattern = r'^(?:[a-zA-Z0-9_](?:[a-zA-Z0-9_-]{0,61}[a-zA-Z0-9_])?\.)*[a-zA-Z0-9_](?:[a-zA-Z0-9_-]{0,61}[a-zA-Z0-9_])?$'
    
    if not re.match(pattern, domain):
        return False, "Invalid domain name format"
    
    return True, ""

def validate_ip(ip: str) -> Tuple[bool, str]:
    """
    Validate IPv4 or IPv6 address
    
    Args:
        ip: IP address to validate
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not ip or not isinstance(ip, str):
        return False, "IP address cannot be empty"
    
    ip = ip.strip()
    
    try:
        ipaddress.ip_address(ip)
        return True, ""
    except ValueError:
        return False, "Invalid IP address format"

def validate_cidr(cidr: str) -> Tuple[bool, str]:
    """
    Validate CIDR notation (e.g., 192.168.1.0/24)
    
    Args:
        cidr: CIDR notation to validate
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not cidr or not isinstance(cidr, str):
        return False, "CIDR notation cannot be empty"
    
    cidr = cidr.strip()
    
    try:
        ipaddress.ip_network(cidr, strict=False)
        return True, ""
    except ValueError as e:
        return False, f"Invalid CIDR notation: {str(e)}"

def validate_targets(targets: List[str]) -> Tuple[bool, str]:
    """
    Validate list of targets (domains, IPs, or CIDR)
    
    Args:
        targets: List of target strings
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not targets:
        return False, "At least one target is required"
    
    if not isinstance(targets, list):
        return False, "Targets must be a list"
    
    if len(targets) > 100:
        return False, "Too many targets (max 100)"
    
    for i, target in enumerate(targets):
        if not isinstance(target, str):
            return False, f"Target {i+1} must be a string"
        
        target = target.strip()
        
        # Try validating as domain, IP, or CIDR
        is_domain, _ = validate_domain(target)
        is_ip, _ = validate_ip(target)
        is_cidr, _ = validate_cidr(target)
        
        if not (is_domain or is_ip or is_cidr):
            return False, f"Target '{target}' is not a valid domain, IP, or CIDR"
    
    return True, ""

def validate_campaign_name(name: str) -> Tuple[bool, str]:
    """
    Validate campaign name
    
    Args:
        name: Campaign name to validate
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not name or not isinstance(name, str):
        return False, "Campaign name cannot be empty"
    
    name = name.strip()
    
    if len(name) < 3:
        return False, "Campaign name must be at least 3 characters"
    
    if len(name) > 100:
        return False, "Campaign name too long (max 100 characters)"
    
    # Allow alphanumeric, spaces, hyphens, underscores
    if not re.match(r'^[a-zA-Z0-9_ -]+$', name):return False, "Campaign name contains invalid characters (only alphanumeric, spaces, hyphens, underscores allowed)"
    
    return True, ""

def validate_username(username: str) -> Tuple[bool, str]:
    """
    Validate username format
    
    Args:
        username: Username to validate
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not username or not isinstance(username, str):
        # Username is optional
        return True, ""
    
    username = username.strip()
    
    if len(username) > 256:
        return False, "Username too long (max 256 characters)"
    
    # Basic username validation (alphanumeric, @, ., -, _)
    if not re.match(r'^[a-zA-Z0-9@._-]+$', username):
        return False, "Username contains invalid characters"
    
    return True, ""

def sanitize_string(value: str, max_length: int = 1000) -> str:
    """
    Sanitize string input by removing/escaping dangerous characters
    
    Args:
        value: String to sanitize
        max_length: Maximum allowed length
        
    Returns:
        Sanitized string
    """
    if not isinstance(value, str):
        return ""
    
    # Remove control characters
    value = re.sub(r'[\x00-\x1f\x7f-\x9f]', '', value)
    
    # Limit length
    value = value[:max_length]
    
    # Strip whitespace
    value = value.strip()
    
    return value

def validate_email(email: str) -> Tuple[bool, str]:
    """
    Validate email address format
    
    Args:
        email: Email address to validate
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not email or not isinstance(email, str):
        # Email is optional
        return True, ""
    
    email = email.strip()
    
    # Basic email regex
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    
    if not re.match(pattern, email):
        return False, "Invalid email address format"
    
    return True, ""

File: web/frontend/src/utils/validation.js
// Frontend validation utilities
import { showValidationError } from "./toast";

/**
 * Validate domain name format
 */
export const validateDomain = (domain) => {
  if (!domain || typeof domain !== "string") {
    return { valid: false, error: "Domain cannot be empty" };
  }

  const trimmed = domain.trim();

  if (trimmed.length > 253) {
    return { valid: false, error: "Domain name too long (max 253 characters)" };
  }

  // Domain regex pattern
  const pattern =
    /^(?:[a-zA-Z0-9_](?:[a-zA-Z0-9_-]{0,61}[a-zA-Z0-9_])?\.)*[a-zA-Z0-9_](?:[a-zA-Z0-9_-]{0,61}[a-zA-Z0-9_])?$/;

  if (!pattern.test(trimmed)) {
    return { valid: false, error: "Invalid domain format" };
  }

  return { valid: true, error: "" };
};

/**
 * Validate IP address (simple check)
 */
export const validateIP = (ip) => {
  if (!ip || typeof ip !== "string") {
    return { valid: false, error: "IP address cannot be empty" };
  }

  const trimmed = ip.trim();

  // Simple IPv4 regex
  const ipv4Pattern = /^(\d{1,3}\.){3}\d{1,3}$/;

  if (!ipv4Pattern.test(trimmed)) {
    return { valid: false, error: "Invalid IP address format" };
  }

  // Check octets are in valid range
  const octets = trimmed.split(".");
  for (const octet of octets) {
    const num = parseInt(octet, 10);
    if (num < 0 || num > 255) {
      return { valid: false, error: "Invalid IP address range" };
    }
  }

  return { valid: true, error: "" };
};

/**
 * Validate CIDR notation
 */
export const validateCIDR = (cidr) => {
  if (!cidr || typeof cidr !== "string") {
    return { valid: false, error: "CIDR notation cannot be empty" };
  }

  const trimmed = cidr.trim();

  // CIDR pattern: IP/mask
  const cidrPattern = /^(\d{1,3}\.){3}\d{1,3}\/\d{1,2}$/;

  if (!cidrPattern.test(trimmed)) {
    return {
      valid: false,
      error: "Invalid CIDR format (use IP/mask like 192.168.1.0/24)",
    };
  }

  const [ip, mask] = trimmed.split("/");

  // Validate IP part
  const ipValidation = validateIP(ip);
  if (!ipValidation.valid) {
    return ipValidation;
  }

  // Validate mask
  const maskNum = parseInt(mask, 10);
  if (maskNum < 0 || maskNum > 32) {
    return { valid: false, error: "Invalid CIDR mask (must be 0-32)" };
  }

  return { valid: true, error: "" };
};

/**
 * Validate campaign name
 */
export const validateCampaignName = (name) => {
  if (!name || typeof name !== "string") {
    return { valid: false, error: "Campaign name cannot be empty" };
  }

  const trimmed = name.trim();

  if (trimmed.length < 3) {
    return {
      valid: false,
      error: "Campaign name must be at least 3 characters",
    };
  }

  if (trimmed.length > 100) {
    return {
      valid: false,
      error: "Campaign name too long (max 100 characters)",
    };
  }

  // Allow alphanumeric, spaces, hyphens, underscores
  if (!/^[a-zA-Z0-9_ -]+$/.test(trimmed)) {
    return { valid: false, error: "Campaign name contains invalid characters" };
  }

  return { valid: true, error: "" };
};

/**
 * Validate target (domain, IP, or CIDR)
 */
export const validateTarget = (target) => {
  if (!target || typeof target !== "string") {
    return { valid: false, error: "Target cannot be empty" };
  }

  const trimmed = target.trim();

  // Try validating as each type
  const domainValidation = validateDomain(trimmed);
  if (domainValidation.valid) return domainValidation;

  const ipValidation = validateIP(trimmed);
  if (ipValidation.valid) return ipValidation;

  const cidrValidation = validateCIDR(trimmed);
  if (cidrValidation.valid) return cidrValidation;

  return {
    valid: false,
    error: "Must be a valid domain, IP address, or CIDR notation",
  };
};

/**
 * Validate email address
 */
export const validateEmail = (email) => {
  if (!email) {
    return { valid: true, error: "" }; // Email is optional
  }

  if (typeof email !== "string") {
    return { valid: false, error: "Invalid email format" };
  }

  const trimmed = email.trim();

  // Basic email regex
  const emailPattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;

  if (!emailPattern.test(trimmed)) {
    return { valid: false, error: "Invalid email address format" };
  }

  return { valid: true, error: "" };
};

/**
 * Validate all campaign form data
 */
export const validateCampaignData = (formData) => {
  const errors = {};

  // Validate campaign name
  const nameValidation = validateCampaignName(formData.name);
  if (!nameValidation.valid) {
    errors.name = nameValidation.error;
  }

  // Validate domain
  const domainValidation = validateDomain(formData.domain);
  if (!domainValidation.valid) {
    errors.domain = domainValidation.error;
  }

  // Validate targets
  if (!formData.targets || !formData.targets.trim()) {
    errors.targets = "At least one target is required";
  } else {
    const targetList = formData.targets.split("\n").filter((t) => t.trim());
    for (let i = 0; i < targetList.length; i++) {
      const targetValidation = validateTarget(targetList[i]);
      if (!targetValidation.valid) {
        errors.targets = `Line ${i + 1}: ${targetValidation.error}`;
        break;
      }
    }
  }

  // Validate optional email
  if (formData.notification_email) {
    const emailValidation = validateEmail(formData.notification_email);
    if (!emailValidation.valid) {
      errors.notification_email = emailValidation.error;
    }
  }

  return {
    valid: Object.keys(errors).length === 0,
    errors,
  };
};
